src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/segmentation/KernelSegmenter.java

```diff
@@ -0,0 +1,523 @@
+        double leftV = leftCost.V;
+        double leftC = leftCost.C;
+
+        double rightD = rightCost.D;
+        final double[] rightW = Arrays.copyOf(rightCost.W, p);
+        double rightV = rightCost.V;
+        double rightC = rightCost.C;
+
+        double totalD = totalCost.D;
+        final double[] totalW = Arrays.copyOf(totalCost.W, p);
+        double totalV = totalCost.V;
+        double totalC = totalCost.C;
+
+        final double[] windowCosts = new double[N];
+        windowCosts[center] = leftC + rightC - totalC;
+
+        double ZdotW;
+        final double windowSizeReciprocal = 1. / windowSize;
+
+        //slide segments along data and use recurrence relations to iteratively update costs
+        for (center = 0; center < N; center++) {
+            final int centerNext = (center + 1) % N;
+            final int endNext = (end + 1) % N;
+
+            //update quantities in left segment
+            leftD -= kernelApproximationDiagonal[start];
+            ZdotW = calculateZdotWAndModifyW(reducedObservationMatrix, p, leftW, start, -1);
+            leftV += -2. * ZdotW + kernelApproximationDiagonal[start];
+
+            leftD += kernelApproximationDiagonal[centerNext];
+            ZdotW = calculateZdotWAndModifyW(reducedObservationMatrix, p, leftW, centerNext, 1);
+            leftV += 2. * ZdotW + kernelApproximationDiagonal[centerNext];
+
+            leftC = leftD - leftV * windowSizeReciprocal;
+
+            //update quantities in right segment
+            rightD -= kernelApproximationDiagonal[centerNext];
+            ZdotW = calculateZdotWAndModifyW(reducedObservationMatrix, p, rightW, centerNext, -1);
+            rightV += -2. * ZdotW + kernelApproximationDiagonal[centerNext];
+
+            rightD += kernelApproximationDiagonal[endNext];
+            ZdotW = calculateZdotWAndModifyW(reducedObservationMatrix, p, rightW, endNext, 1);
+            rightV += 2. * ZdotW + kernelApproximationDiagonal[endNext];
+
+            rightC = rightD - rightV * windowSizeReciprocal;
+
+            //update quantities in total segment
+            totalD -= kernelApproximationDiagonal[start];
+            ZdotW = calculateZdotWAndModifyW(reducedObservationMatrix, p, totalW, start, -1);
+            totalV += -2. * ZdotW + kernelApproximationDiagonal[start];
+
+            totalD += kernelApproximationDiagonal[endNext];
+            ZdotW = calculateZdotWAndModifyW(reducedObservationMatrix, p, totalW, endNext, 1);
+            totalV += 2. * ZdotW + kernelApproximationDiagonal[endNext];
+
+            totalC = totalD - 0.5 * totalV * windowSizeReciprocal;
+
+            //record cost of changepoint at this position
+            windowCosts[centerNext] = leftC + rightC - totalC;
+
+            //slide windows
+            start = (start + 1) % N;
+            end = endNext;
+        }
+        return windowCosts;
+    }
+
+    private static double calculateZdotWAndModifyW(final RealMatrix reducedObservationMatrix,
```

Comment:
Could you add a brief description of the `signMultiplier`, and what does _modifying_ `W` means?

Reply:
In my opinion, extracting `calculateZdotWAndModifyW` has made the code less readable :)

Reply:
Actually, I think I agree...I'll change it back if @davidbenjamin doesn't mind.  (snip snap snip snap!)